libeos-payg/util.c

```diff
@@ -0,0 +1,247 @@
+    return debug_flags;
+
+  /* For now we only look at the first byte but let's allow it to be bigger so
+   * it can be extended in the future.
+   */
+  ret = efi_get_variable (EOSPAYG_GUID, "EOSPAYG_debug", &debug_efivar, &data_size, &attributes);
+  if (ret < 0 || !debug_efivar || data_size < 1)
+    {
+      g_printerr ("%s: Failed to read EOSPAYG_debug\n", G_STRFUNC);
+      return debug_flags;
+    }
+
+  g_debug ("%s: EOSPAYG_debug is set to %u", G_STRFUNC, debug_efivar[0]);
+
+  /* The C standard doesn't guarantee the underlying type of an enum, so let's
+   * convert manually to be safe.
+   */
+  if (debug_efivar[0] & EPG_DEBUG_SET_G_MESSAGES_DEBUG_ALL)
+    debug_flags |= EPG_DEBUG_SET_G_MESSAGES_DEBUG_ALL;
+  if (debug_efivar[0] & EPG_DEBUG_SECURE_BOOT_OFF)
+    debug_flags |= EPG_DEBUG_SECURE_BOOT_OFF;
+  if (debug_efivar[0] & EPG_DEBUG_SECURE_BOOT_ON)
+    debug_flags |= EPG_DEBUG_SECURE_BOOT_ON;
+  if (debug_efivar[0] & EPG_DEBUG_EOSPAYG_ACTIVE_OFF)
+    debug_flags |= EPG_DEBUG_EOSPAYG_ACTIVE_OFF;
+  if (debug_efivar[0] & EPG_DEBUG_EOSPAYG_ACTIVE_ON)
+    debug_flags |= EPG_DEBUG_EOSPAYG_ACTIVE_ON;
+  if (debug_efivar[0] & EPG_DEBUG_DONT_USE_WATCHDOG)
+    debug_flags |= EPG_DEBUG_DONT_USE_WATCHDOG;
+  if (debug_efivar[0] & EPG_DEBUG_DONT_USE_LSM)
+    debug_flags |= EPG_DEBUG_DONT_USE_LSM;
+  if (debug_efivar[0] & EPG_DEBUG_SKIP_SECURITYLEVEL_CHECK)
+    debug_flags |= EPG_DEBUG_SKIP_SECURITYLEVEL_CHECK;
+
+  return debug_flags;
+}
+
+/**
+ * epg_set_debug_env_vars:
+ *
+ * If the EOSPAYG_debug EFI variable has the bit set for
+ * EPG_DEBUG_SET_G_MESSAGES_DEBUG_ALL, this function sets the environment
+ * variable "G_MESSAGES_DEBUG" to the value "all" (unless G_MESSAGES_DEBUG is
+ * already set).
+ *
+ * This may be extended in the future to include other environment variables.
+ */
+void
+epg_set_debug_env_vars (void)
+{
+  EpgDebugFlags debug_flags = _read_eospayg_debug ();
+
+  if (debug_flags & EPG_DEBUG_SET_G_MESSAGES_DEBUG_ALL)
+    g_setenv ("G_MESSAGES_DEBUG", "all", FALSE);
+}
+
+/**
+ * epg_get_secure_boot_enabled:
+ *
+ * If the EOSPAYG_debug EFI variable has the bit set for
+ * EPG_DEBUG_SECURE_BOOT_OFF or EPG_DEBUG_SECURE_BOOT_ON, this will return
+ * %FALSE or %TRUE respectively, regardless of whether secure boot is actually
+ * enabled. Otherwise, the SecureBoot EFI variable is checked to determine if
+ * the current boot is secure.
+ *
+ * Returns: %TRUE if eos-paygd should act as if Secure Boot is enabled, and
+ *   %FALSE otherwise
+ */
+gboolean
+epg_get_secure_boot_enabled (void)
+{
+  EpgDebugFlags debug_flags = _read_eospayg_debug ();
+  uint8_t *secboot = NULL;
+  size_t data_size = 0;
+  uint32_t attributes;
+  int ret;
+
+  if (debug_flags & EPG_DEBUG_SECURE_BOOT_OFF &&
+      debug_flags & EPG_DEBUG_SECURE_BOOT_ON)
+    g_printerr ("%s: Both EPG_DEBUG_SECURE_BOOT_OFF and EPG_DEBUG_SECURE_BOOT_ON are set\n", G_STRFUNC);
+
+  if (debug_flags & EPG_DEBUG_SECURE_BOOT_OFF)
```

Comment:
Optional pedantry - would it be useful to know if our override didn't actually override state?
ie) secure boot was already off but we forced it off

Reply:
That would require reading EFI variables just for the sake of a debug message. I think that's not worth it, in case there's a price to pay in terms of performance or wear on the flash memory.